File: lab 7/java/src/Philosopher.java
import java.util.Random;

public abstract class Philosopher extends Thread
{
    Random rand = new Random();
    int number;
    Fork f1;
    Fork f2;
    long waitingTime = 0;
    long waitingCount = 0;
    long startedWaiting;
    long finishedWaiting;

    public void run()
    {
        try
        {
            for(int i = 0; i < 100; i++)
            {
                think();
                startedWaiting = System.currentTimeMillis();
                eat();
            }
        }
        catch(InterruptedException e)
        {
            e.printStackTrace();
        }
    }

    private void think() throws InterruptedException
    {
        sleep(Math.abs(rand.nextInt()) % 100);
    }

    abstract protected void eat()  throws InterruptedException;

    protected void consume(Fork f1, Fork f2) throws InterruptedException
    {
        f1.take();
        sleep(1);
        f2.take();
        finishedWaiting = System.currentTimeMillis();
        waitingCount++;
        waitingTime = waitingTime + finishedWaiting - startedWaiting;
        sleep(Math.abs(rand.nextInt()) % 100);
        f1.putBack();
        f2.putBack();
    }
}

File: lab 7/scala/src/main/scala/Philosopher.scala
import java.util.Random

abstract class Philosopher(number: Int, f1: Fork, f2: Fork) extends Thread {
  val rand: Random = new Random
  var waitingTime = 0
  var waitingCount = 0
  var startedWaiting = 0
  var finishedWaiting = 0

  override def run(): Unit = {
    try {
      var i: Int = 0
      while ( {
        i < 100
      }) {
        think()
        startedWaiting = System.currentTimeMillis.toInt
        eat()

        {
          i += 1; i - 1
        }
      }
    } catch {
      case e: InterruptedException =>
        e.printStackTrace()
    }
  }

  @throws[InterruptedException]
  private def think(): Unit = {
    Thread.sleep(Math.abs(rand.nextInt) % 100)
  }

  @throws[InterruptedException]
  protected def eat(): Unit

  @throws[InterruptedException]
  protected def consume(f1: Fork, f2: Fork): Unit = {
    f1.take()
    Thread.sleep(1)
    f2.take()
    finishedWaiting = System.currentTimeMillis.toInt
    waitingCount += 1
    waitingTime = waitingTime + finishedWaiting - startedWaiting
    Thread.sleep(Math.abs(rand.nextInt) % 100)
    f1.putBack()
    f2.putBack()
  }
}

